lib/puppet/provider/user/windows_adsi.rb

```diff
@@ -123,16 +307,82 @@ def password
     user.password_is?( @resource[:password] ) ? @resource[:password] : nil
   end
 
+  # Sets the user's password.
+  #
+  # @api private
+  def set_password(password, existing_user)
+    if existing_user
+      # We always want to set a new user's password. However for
+      # an existing user, we need to check if changing their
+      # password makes sense
+
+      # Do NOT attempt to retrieve a new user's attributes. Doing so
+      # will raise an exception because the user's ADSI property values
+      # (some of which correspond to their attributes) are only loaded
+      # into the property cache _after_ they've been committed.
+      current_attributes = attributes
+
+      if current_attributes[:password_change_required] == 'true'
+        warning _("The account '%{name}' requires the user to change their password on the next sign-in; Puppet will not reset the password. If you still wish to reset the user's password, then set password_change_required to false in the attributes property and re-run Puppet.") % { name: name }
+        return
+      end
+  
+      if current_attributes[:disabled] == 'true'
+        warning _("The user account '%s' is disabled; Puppet will not reset the password. If you still wish to reset the user's password, then set disabled to false in the attributes property and re-run Puppet." % @resource[:name])
+        return
+      end
+  
+      if user.locked_out?
+        warning _("The user account '%s' is locked out; Puppet will not reset the password" % @resource[:name])
+        return
+      end
+  
+      if user.expired?
+        warning _("The user account '%s' is expired; Puppet will not reset the password" % @resource[:name])
+        return
+      end
+    end
+
+    user.password = password
+
+    # The user's been committed. Thus, it is safe to retrieve
+    # their attributes.
+
+    if (attributes_prop = @resource.property(:attributes))
+      specified_attributes = attributes_prop.hashify_should
+      unless specified_attributes.key?(:password_never_expires)
+        warning _("The attributes property did not specify the 'password_never_expires' attribute, which is currently set to %{password_never_expires} on the system. If this is undesirable behavior, then please specify a value for the 'password_never_expires' attribute, and re-run Puppet.") % { password_never_expires: attributes[:password_never_expires] }
```

Comment:
Should this warning message be printed? I added it here because we always set the password to never expire, so I figured it would be useful to let the user know that they should probably manage that attribute whenever they're setting passwords.